jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3640,32 +3711,52 @@ private static boolean counterPredicate(int endValue, int counter) {
 	public static MethodHandle iteratedLoop(MethodHandle iteratorHandle, MethodHandle initHandle, MethodHandle bodyHandle) throws NullPointerException, IllegalArgumentException {
 		validateArgumentsOfIteratedLoop(iteratorHandle, initHandle, bodyHandle);
 		
-		/* Method handles to Iterator.hasNext(), Iterator.next() and Iterable.iterator() are required
-		 * to construct a clause array in the the generic loop.
-		 */
-		MethodHandle iteratorNextElement = null;
-		MethodHandle iteratorHasNextElement = null;
-		try {
-			iteratorNextElement = Lookup.internalPrivilegedLookup.findVirtual(Iterator.class, "next", MethodType.methodType(Object.class)); //$NON-NLS-1$
-			iteratorHasNextElement = Lookup.internalPrivilegedLookup.findVirtual(Iterator.class, "hasNext", MethodType.methodType(boolean.class)); //$NON-NLS-1$
-		} catch (IllegalAccessException | NoSuchMethodException e) {
-			throw new InternalError("The next/hasNext method doesn't exist or it fails in the access checking", e); //$NON-NLS-1$
-		}
+		MethodType bodyType = bodyHandle.type;
+		Class<?> bodyReturnType = bodyType.returnType;
 		
 		/* The init handle for iterator is set to the method handle to Iterable.iterator()
-		 * by default if it is null.
+		 * by default if it is null. The default iterator handle parameters are adjusted 
+		 * to accept the leading A type.
 		 */
 		MethodHandle initIterator = iteratorHandle;
 		if (null == initIterator) {
+			int bodyParamTypesLength = bodyHandle.type.arguments.length;
+			int bodyIterationVarLength = 2;
+			if (void.class == bodyReturnType) {
+				bodyIterationVarLength = 1;
+			}
+			
+			/* If body has an A type, the default iterator handle is adjusted to accept the leading A type. 
+			 * Validation has already confirmed that the leading A type is an Iterable or subtype thereof.
+			 */
+			Class<?> iterableType;
```

Comment:
The logic is clearer if it defaults to `Iterable.class` as a reader only needs to look at 1 branch rather than 2.

```
Class<?> iterableType = Iterable.class;`
if (bodyParamTypesLength > bodyIterationVarLength) {
 		iterableType = bodyHandle.type.arguments[bodyIterationVarLength];
}
```

Reply:
@theresa-m , please update against the suggestion above.

Reply:
Just updated this.